pages/en/zkapps/advanced-concepts.mdx

```diff
@@ -57,8 +57,102 @@ This is all you need to know about events! Think of them as a convenience featur
 
 ```ts
 class SmartContract {
-  static events: Record<string, any>;
+  static events?: Record<string, any>;
 
   emitEvent(name: string, event: any): void;
 }
 ```
+
+## Actions and Reducer
+
+<Alert kind="info">Experimental. This API may change.</Alert>
+
+Like events, **actions** are arbitrary information passed along with a zkApp transaction. However, actions give you an additional power: you can process previous actions in a smart contract! Under the hood, this is possible because we store a commitment to the history of dispatched actions on every account -- the **actionsHash**. It allows us to prove that the actions you process are, in fact, the actions that were dispatched to the same smart contract.
+
+Using actions and a "lagging state" pattern, you can write zkApps that can _process concurrent state updates by multiple users_ -- see the next section. Besides that, we imagine all kinds of use cases where actions act as a built-in, "append-only" off-chain storage layer.
```

Comment:
Could it be nice to have a paragraph describing how this is an adaptation of existing "Redux" / "Elm Architecture" models (see my original explanation here https://github.com/o1-labs/snarkyjs/issues/265#issuecomment-1178307804 )?